File: ts/test/diff.test.ts
/* Copyright (c) 2025 Richard Rodger, MIT License */

// The diff API around the shared rows (G7 phase 6). Every change kind
// and every path shape is pinned by test/spec/diff.tsv in both ports;
// what is left here is the options, which no row exercises.

import { describe, test } from 'node:test'
import * as Assert from 'node:assert'
import * as Fs from 'node:fs'
import * as Os from 'node:os'
import * as Path from 'node:path'

import { diff } from '../dist/aontu'


describe('diff', () => {

  // A change at the ROOT itself: two documents that are not both maps
  // (or both lists) compare as whole values, at `$`.
  test('root-change-is-reported-at-the-root', () => {
    const r = diff('a: 1', '[1]')
    Assert.equal(r.ok, true)
    Assert.deepEqual(r.changes.map((c) => c.path), ['$'])
    Assert.equal(r.changes[0].kind, 'changed')
  })

  test('each-side-resolves-includes-from-its-own-directory', () => {
    const left = Fs.mkdtempSync(Path.join(Os.tmpdir(), 'aontu-diff-l-'))
    const right = Fs.mkdtempSync(Path.join(Os.tmpdir(), 'aontu-diff-r-'))
    Fs.writeFileSync(Path.join(left, 'part.aon'), 'k: 1')
    Fs.writeFileSync(Path.join(right, 'part.aon'), 'k: 2')
    // The entry file has to EXIST: the resolver stats it to root the
    // relative load, exactly as it does for `aontu <file>`.
    Fs.writeFileSync(Path.join(left, 'doc.aon'), 'a: @"part.aon"')
    Fs.writeFileSync(Path.join(right, 'doc.aon'), 'a: @"part.aon"')

    const r = diff('a: @"part.aon"', 'a: @"part.aon"', {
      leftPath: Path.join(left, 'doc.aon'),
      rightPath: Path.join(right, 'doc.aon'),
    })
    Assert.equal(r.ok, true)
    Assert.deepEqual(r.changes,
      [{ kind: 'changed', left: '1', path: '$.a.k', right: '2' }])
  })

})

File: go/diff_test.go
/* Copyright (c) 2025 Richard Rodger, MIT License */

// The diff API around the shared rows (G7 phase 6). Every change kind
// is pinned by test/spec/diff.tsv in both ports; what is left here is
// the per-document base path, which no row exercises and which the
// cross-package CLI runs do not count toward this package's coverage.

package aontu

import (
	"os"
	"path/filepath"
	"testing"
)

func TestDiffResolvesIncludesFromEachSidesBase(t *testing.T) {
	left := t.TempDir()
	right := t.TempDir()
	for dir, k := range map[string]string{left: "k: 1", right: "k: 2"} {
		if err := os.WriteFile(
			filepath.Join(dir, "part.aon"), []byte(k), 0o600); nil != err {
			t.Fatal(err)
		}
	}

	r := Diff(`a: @"part.aon"`, `a: @"part.aon"`, &DiffOptions{
		LeftPath:  left,
		RightPath: right,
	})
	if !r.OK || 1 != len(r.Changes) {
		t.Fatalf("bad report: %+v", r)
	}
	c := r.Changes[0]
	if "$.a.k" != c.Path || "1" != c.Left || "2" != c.Right {
		t.Fatalf("bad change: %+v", c)
	}
}
